core/client/src/client.rs

```diff
@@ -579,6 +594,74 @@ impl<B, E, Block> Client<B, E, Block> where
 		let info = self.backend.blockchain().info().map_err(|e| error::Error::from_blockchain(Box::new(e)))?;
 		Ok(self.header(&BlockId::Hash(info.best_hash))?.expect("Best block header must always exist"))
 	}
+
+	/// Get the most recent block hash of the best (longest) chains
+	/// that contain block with the given `target_hash`.
+	/// If `maybe_max_block_number` is `Some(max_block_number)`
+	/// the operation is performed as if there are no blocks greater `max_block_number`
+	/// ignoring all blocks with numbers greater
+	/// TODO [snd] possibly implement this on blockchain::Backend and just redirect here
+	pub fn best_containing(&self, target_hash: Block::Hash, maybe_max_number: Option<<<Block as BlockT>::Header as HeaderT>::Number>) -> error::Result<Option<Block::Hash>> {
+		let target_header = {
+			match self.backend.blockchain().header(BlockId::Hash(target_hash))? {
+				Some(x) => x,
+				// target not in blockchain
+				None => { return Ok(None); },
+			}
+		};
+
+		if let Some(max_number) = maybe_max_number {
+			// target outside search range
+			if target_header.number() > &max_number {
+				return Ok(None);
+			}
+		}
+
+		// for each chain. longest chain first. shortest last
+		for leaf_hash in self.backend.blockchain().leaves()? {
```

Comment:
this loop could be pre-empted by checking if the block with number `target_header.number()` in the "canonical" chain is `target_header`. Could return the "best block" in that case, or its ancestor with `max_number`. Although the import lock probably has to be held for safety with that approach, to ensure that a block import doesn't race against this thread and trigger a reorganization.

Reply:
i believe i tried this before by checking whether the hash of the block at `target_header.number()` matches `target_hash`: https://github.com/paritytech/substrate/blob/c051949f40db9dff5bcab7efa50988464fa10a0a/core/client/src/client.rs#L593

however `self.backend.blockchain().hash(*target_header.number())` returned `None`
for those numbers that were not uniquely mappable to a hash i.e. there were forks and multiple hashes for that number. i'll investigate again. maybe there's a better method.

Reply:
he current blockchain backend doesn't handle forks at all, so I've been rewriting it to do so in a local branch.

Reply:
> this loop could be pre-empted by checking if the block with number target_header.number() in the "canonical" chain is target_header

how do i get the header/hash that's in the canonical chain at position `number`?

i assumed i could do this through `...blockchain().header(BlockId::Number(number))`  / `...blockchain().hash(number)`.

**both do not seem to return the header/hash in the canonical chain but rather the header/hash that was last imported at that position.**

i've added a test exposing that problem here https://github.com/paritytech/substrate/pull/740/commits/57e758b48003472fa31f64ec49e05a471025cda1. note that it only tests the in-memory backend.

is this indeed a bug that should be fixed? is there another method to retrieve a block given a number in the canonical chain? should i leave out the pre-emption for now?

Reply:
> both do not seem to return the header/hash in the canonical chain but rather the header/hash that was last imported at that position.

This is exactly what  #760 is meant to fix; the underlying blockchain DB literally could not accommodate forks before.

Reply:
perfect! this should work out then

Reply:
pre-emption is in. the tests should pass once #760 is merged and this is rebased on it